src/audio/multiband_drc/multiband_drc.c

```diff
@@ -0,0 +1,605 @@
+		ret = comp_data_blob_set_cmd(cd->model_handler, cdata);
+		break;
+	default:
+		comp_err(dev, "multiband_drc_cmd_set_data() error: invalid cdata->cmd");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+static int multiband_drc_cmd(struct comp_dev *dev, int cmd, void *data, int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+	comp_info(dev, "multiband_drc_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = multiband_drc_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = multiband_drc_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	default:
+		comp_err(dev, "multiband_drc_cmd(), invalid command");
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static int multiband_drc_trigger(struct comp_dev *dev, int cmd)
+{
+	int ret;
+
+	comp_info(dev, "multiband_drc_trigger()");
+
+	ret = comp_set_state(dev, cmd);
+	if (ret == COMP_STATUS_STATE_ALREADY_SET)
+		ret = PPL_STATUS_PATH_STOP;
+
+	return ret;
+}
+
+static void multiband_drc_process(struct comp_dev *dev, struct comp_buffer *source,
+				  struct comp_buffer *sink, int frames,
+				  uint32_t source_bytes, uint32_t sink_bytes)
+{
+	struct multiband_drc_comp_data *cd = comp_get_drvdata(dev);
+
+	buffer_invalidate(source, source_bytes);
+
+	cd->multiband_drc_func(dev, &source->stream, &sink->stream, frames);
+
+	buffer_writeback(sink, sink_bytes);
+
+	/* calc new free and available */
+	comp_update_buffer_consume(source, source_bytes);
+	comp_update_buffer_produce(sink, sink_bytes);
+}
+
+static int multiband_drc_copy(struct comp_dev *dev)
+{
+	struct comp_copy_limits cl;
+	struct multiband_drc_comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sourceb;
+	struct comp_buffer *sinkb;
+	int ret;
+
+	comp_dbg(dev, "multiband_drc_copy()");
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
+				source_list);
+
+	/* Check for changed configuration */
+	if (comp_is_new_data_blob_available(cd->model_handler)) {
+		cd->config = comp_get_data_blob(cd->model_handler, NULL, NULL);
+		ret = multiband_drc_setup(cd, (int16_t)sourceb->stream.channels,
+					  sourceb->stream.rate);
+		if (ret < 0) {
+			comp_err(dev, "multiband_drc_copy(), failed DRC setup");
+			return ret;
+		}
+	}
+
+	/* Get source, sink, number of frames etc. to process. */
+	comp_get_copy_limits_with_lock(sourceb, sinkb, &cl);
+
+	/* Run Multiband DRC function */
+	multiband_drc_process(dev, sourceb, sinkb, cl.frames, cl.source_bytes, cl.sink_bytes);
+
+	return 0;
+}
+
+static int multiband_drc_prepare(struct comp_dev *dev)
+{
+	struct multiband_drc_comp_data *cd = comp_get_drvdata(dev);
+	struct sof_ipc_comp_config *config = dev_comp_config(dev);
+	struct comp_buffer *sourceb;
+	struct comp_buffer *sinkb;
+	uint32_t sink_period_bytes;
+	int ret;
+
+	comp_info(dev, "multiband_drc_prepare()");
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
```

Comment:
shouldnt this be the last step?

Reply:
For me it seems fine to call here, we could return earlier if comp_set_state() returns error.
@singalsu what do you think?

Reply:
but we won't stay in current state if we fail either

Reply:
@johnylin76 ping on this

Reply:
I still think it should be done as it is:
1. we could return earlier if comp_set_state() returns error
2. if we fail on preparing, then "goto err" will replace the state by COMP_TRIGGER_RESET, we won't stay in PREPARE state.